File: LEETCODE/Greedy/2943_maximumAreaOfSquareHoleInAGrid/2943_maximumAreaOfSquareHoleInAGrid.java
class Solution {
    private int findLongestConsecutive(int A[]){
        int n= A.length;
        int currMax=1,maxi=1;

        for(int i=1;i<n;i++){
            if(A[i]-A[i-1]==1){
                currMax++;
            }
            else
                currMax=1;
            maxi= Math.max(maxi,currMax);
        }
        return maxi;
    }
    public int maximizeSquareHoleArea(int n, int m, int[] hBars, int[] vBars) {  /*
          PROBLEM KA CORE IDEA:
          ->    Agar k horizontal bars consecutively remove hue,
                toh vertical direction me size = k + 1
          ->    Same for vertical bars.
          ->    Square banane ke liye:
                side = min(horizontal_gap, vertical_gap)
         */

        // Step 0:
        // Bars ko sort karna compulsory
        // Warna consecutive logic ka koi matlab hi nahi

        Arrays.sort(hBars);
        Arrays.sort(vBars);

        //Step 1: find the longest consecutive subarrays in both cases
        int longestConsecutiveHBar= findLongestConsecutive(hBars);
        int longestConsecutiveVBar= findLongestConsecutive(vBars);
           /*
          Step 2 (MOST IMPORTANT):
         
          Agar 3 bars consecutively remove hue,
          toh actual gap = 3 + 1 = 4
         
          Square ka side hamesha MIN hoga
          (square bhai rectangle nahi hota)
         */

        int side= Math.min(longestConsecutiveVBar,longestConsecutiveHBar)+1;
        return side*side;  
    }
}

File: LEETCODE/Greedy/2943_maximumAreaOfSquareHoleInAGrid/2943_maximumAreaOfSquareHoleInAGrid.py
    class Solution:
        def maximizeSquareHoleArea(self, n: int, m: int, hBars: List[int], vBars: List[int]) -> int:
            
            def longest_consecutive(arr: List[int])->int:
                arr.sort()
                best=curr=1

                for i in range(1,len(arr)):
                    if arr[i] == arr[i-1]+1:
                        curr+=1
                    else:
                        curr=1
                    best= max(curr,best)
                
                return best
            
            side= min(
                longest_consecutive(hBars),
                longest_consecutive(vBars)
            )+1

            return side*side